packages/devtools-connection/src/utils/event-emitter.js

```diff
@@ -2,138 +2,259 @@
+        // in another function.
+        // So we iterate all the listeners to check if any of them is a wrapper to
+        // the `listener` given.
+        for (const value of listenersForType.values()) {
+          if (onceOriginalListener in value && value[onceOriginalListener] === listener) {
+            listenersForType.delete(value);
+            break;
+          }
+        }
+      }
+    } else if (length === 2) {
+      // No listener was given, it means we're removing all the listeners from
+      // the given event's `type`.
+      if (events.has(type)) {
+        events.delete(type);
+      }
+    } else if (length === 1) {
+      // With only the `target` given, we're removing all the isteners from the object.
+      events.clear();
+    }
+  }
 
   /**
-   * Remove a previously-registered event listener.  Works for events
-   * registered with either on or once.
+   * Registers an event `listener` that is called only the next time an event
+   * of the specified `type` is emitted on the given event `target`.
+   * It returns a promised resolved once the specified event `type` is emitted.
    *
-   * @param string aEvent
-   *        The event name whose listener we're disconnecting.
-   * @param function aListener
-   *        The listener to remove.
+   * @param {Object} target
+   *    Event target object.
+   * @param {String} type
+   *    The type of the event.
+   * @param {Function|Object} [listener]
+   *    The listener that processes the event.
+   * @return {Promise}
+   *    The promise resolved once the event `type` is emitted.
    */
-  off: function EventEmitter_off(aEvent, aListener) {
-    if (!this._eventEmitterListeners) {
+  static once(target, type, listener) {
+    return new Promise(resolve => {
+      // This is the actual listener that will be added to the target's listener, it wraps
+      // the call to the original `listener` given.
+      const newListener = (first, ...rest) => {
+        // To prevent side effects we're removing the listener upfront.
+        EventEmitter.off(target, type, newListener);
+
+        if (listener) {
+          if (isEventHandler(listener)) {
+            // if the `listener` given is actually an object that handles the events
+            // using `EventEmitter.handler`, we want to call that function, passing also
+            // the event's type as first argument, and the `listener` (the object) as
+            // contextual object.
+            listener[handler](type, first, ...rest);
+          } else {
+            // Otherwise we'll just call it
+            listener.call(target, first, ...rest);
+          }
+        }
+
+        // We resolve the promise once the listener is called.
+        resolve(first);
+      };
+
+      newListener[onceOriginalListener] = listener;
+      EventEmitter.on(target, type, newListener);
+    });
+  }
+
+  static emit(target, type, ...rest) {
+    if (!(eventListeners in target)) {
```

Comment:
Logging is disabled by default, but I'm wondering if it makes sense to keep it in. The code doesn't look too intrusive. Would make keeping this file in sync easier, as then it is almost a direct copy. The only thing that stands out in that case is services. If we decide to keep it out, lets add to the comment that logging has been removed.